### ISSUE TITLE
refactoring sat-api packages

### ISSUE THREAD

**Issue by matthewhanson** [sat-utils/sat-api#67]
Now that the initial work in packaging up sat-api as a series of modules has been done thanks to @scisco it's time to start thinking about what the core group of packages should look like.

Right now we have:
- api: lambda function for use with API Gateway as the backend, includes all the other packages
- api-lib: a collection of modules for ingesting into and searching elasticsearch, and ingesting CSV files
- manager: lambda function for running management functions on elasticsearch
- ingest: lambda function that fetches a CSV file, splits it up, and calls a specified lambda function on each one to ingest it into elasticsearch
- landsat: the landsat ingesting lambda called on each split CSV
- sentinel: the sentinel ingesting lambda called on each split CSV

However, ingesting CSV files, while a stopgap for now, is not the long term preferred method of ingesting items. Instead we will want a method to ingest a STAC catalog by providing a URL to any STAC json file. It would then traverse the tree down ingesting all records. Since the catalog is STAC there is no need for sensor specific logic here.

The catalog ingestor can be used to ingest a catalog to populate sat-api with existing data but the question remains how to update it. How do we know what new records to ingest moving forward? Part of that answer is an SNS ingestor. But what happens if the service goes down and some records are missing. We also need a way to backfill sat-api in this case, and I'm not sure how we efficiently pick up the STAC catalog tree again. How can we go to the last ingested record(s) and discover what is new?

So my proposed new packages would be:

### elasticsearch

The backend library for using elasticsearch, both adding records (pass in a STAC item), and searching it.

### postgis

A postgis-based backend library for adding records and searching

### ingest
Ingests a catalog.json file, traversing it and adding records. Run as Lambda (?) or as an ECS task.  User selectable backend (e.g., ES or PostGIS).

### sns-ingest
Lambda function for ingesting SNS messages. User selectable backend (e.g., ES or PostGIS)

### api
Lambda function backend to an API that uses a selectable backend (e.g., ES or PostGIS)

### manager (?)
I'm not sure if this this is even needed anymore, it's a matter of convenience, we could opt to have management scripts that can be run locally and manage AWS resources. This may be a bit more transparent.

cc @fredliporace

**Comment by fredliporace** [comment]
@matthewhanson things will be clear for me when I get the sns-ingest module working, but my view now is that it is possible to use the same SNS/SQS infrastructure that is used to update a catalog to ingest a new one. For the update the SQS is populated with new data as soon as it is notified through SNS, and in the 'ingestion' scenario the SQS could be populated while the catalog is traversed. Once the messages (stac items) are in the queue it is possible to use lambda to process the items, and the scaling is automatic. Since the messages are only removed from the queue when the lambda function is successfully executed the method is also robust.

**Comment by matthewhanson** [comment]
Makes sense @fredliporace, much of the core of the code probably can go in a library function.

I like the idea of ingesting by using SQS, a Kinesis stream may also be an option.

**Comment by fredliporace** [comment]
@matthewhanson see https://aws.amazon.com/kinesis/data-streams/faqs/ for a comparison of Kinesis and SQS, and when to use one or the other. I have no kinesis experience but from what I read it seems that SQS is best suited for our case.

I've deployed a sat-api instance with a sns feed receiving CBERS MUX and AWFI metadata as they are ingested. Will post my view on how it is working and some potential issues here. One thing I noticed is that if the feed is 'slow' the lambda trigger is fired one time for each stac item, I think it would be better to wait some time and fire one lambda call for the configured batch size, 10 in my case.

**Comment by fredliporace** [comment]
I developed a crawler to send all static STAC items to a SNS topic and have them ingested by sat-api, the code is [here](https://github.com/fredliporace/stac-crawler). This may be checked as a reference to the proposed 'ingest' module. Tested with a static catalog of 9437 items and it worked fine. Next step  is to try with the complete CBERS-4 catalog on AWS, I expect that some work is still needed to improve reliability under heavy load.
